pkg/pillar/cmd/zedagent/attesttask.go

```diff
@@ -78,9 +291,108 @@ func attestModuleStart(ctx *zedagentContext) error {
 	if ctx.attestCtx == nil {
 		return fmt.Errorf("No attest module context")
 	}
-	if ctx.attestCtx.attestModCtx == nil {
+	if ctx.attestCtx.attestFsmCtx == nil {
 		return fmt.Errorf("No state machine context found")
 	}
-	go ctx.attestCtx.attestModCtx.EnterEventLoop()
+	go ctx.attestCtx.attestFsmCtx.EnterEventLoop()
+	zattest.Kickstart(ctx.attestCtx.attestFsmCtx)
 	return nil
 }
+
+// pubsub functions
+func handleAttestQuoteModify(ctxArg interface{}, key string, quoteArg interface{}) {
+
+	//Store quote received in state machine
+	ctx, ok := ctxArg.(*zedagentContext)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected ctx type %T", ctxArg)
+	}
+
+	quote, ok := quoteArg.(types.AttestQuote)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected pub type %T", quoteArg)
+	}
+
+	if ctx.attestCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to attestCtx")
+	}
+
+	//Deepcopy quote into InternalQuote
+	attestCtx := ctx.attestCtx
+	attestCtx.InternalQuote = &types.AttestQuote{}
+	buf, _ := json.Marshal(&quote)
+	json.Unmarshal(buf, attestCtx.InternalQuote)
+
+	if attestCtx.attestFsmCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to attestFsmCtx")
+	}
+	//Trigger event on the state machine
+	zattest.InternalQuoteRecvd(attestCtx.attestFsmCtx)
+
+	log.Infof("handleAttestQuoteModify done for %s", quote.Key())
+	return
+}
+
+func handleAttestQuoteDelete(ctxArg interface{}, key string, quoteArg interface{}) {
+	//Delete quote received in state machine
+
+	ctx, ok := ctxArg.(*zedagentContext)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected ctx type %T", ctxArg)
+	}
+
+	quote, ok := quoteArg.(types.AttestQuote)
+	if !ok {
+		log.Fatalf("[ATTEST] Unexpected pub type %T", quoteArg)
+	}
+
+	if ctx.attestCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to attestCtx")
+	}
+
+	attestCtx := ctx.attestCtx
+	if attestCtx.InternalQuote == nil {
+		log.Warnf("[ATTEST] Delete received while InternalQuote is unpopulated, ignoring")
+		return
+	}
+
+	if attestCtx.attestFsmCtx == nil {
+		log.Fatalf("[ATTEST] Uninitialized access to attestFsmCtx")
+	}
+
+	if reflect.DeepEqual(quote.Nonce, attestCtx.InternalQuote.Nonce) {
+		attestCtx.InternalQuote = nil
+	} else {
+		log.Warnf("[ATTEST] Nonce didn't match, ignoring incoming delete")
+	}
+	log.Infof("handleAttestQuoteDelete done for %s", quote.Key())
+	return
+}
+
+func publishAttestNonce(ctx *attestContext) {
+	nonce := types.AttestNonce{
+		Nonce:     ctx.Nonce,
+		Requester: agentName,
+	}
+	key := nonce.Key()
+	log.Debugf("[ATTEST] publishAttestNonce %s", key)
+	pub := ctx.pubAttestNonce
+	pub.Publish(key, nonce)
+	log.Debugf("[ATTEST] publishAttestNonce done for %s", key)
+}
+
+func unpublishAttestNonce(ctx *attestContext) {
```

Comment:
Given that the Key for the AttestNonce is the nonce itself we can potentially have multiple nonces published.
How do we ensure that we always unpublish the old?
Would it make sense to do a GetAll and fatal if it returns something after we've done the unpublish?

Reply:
Agreed, and updated the code accordingly